Dynamic spinup function

pat-schmitt opened the pull request:
Here is my updated version of the dynamic spinup function.
The main changes are:

- safeguards included which prevent taking too large steps between the minimisation iterations (also changed from a polyfit to a splinfit = a linear interpolation)
- errors that can occur during minimisation are handled explicitly now (e.g. glacier out of domain, or glacier ice free after spinup but still too large at the end) 
- include `ignore_error` option, if an error occurs the glacier is stored without a dynamic spinup (so they are available for the following tasks)
- included the dynamic spinup function in `run_prepro_levels`

Things to do:
- [ ] the final test of current implementation (currently in queue on cluster, but no big changes between the current version and the last successfully tested version)
- [ ] define additional glaciers for more advanced testing, of the handling of all error cases that can occur during the dynamic spinup 
- I am currently adding quite some variables to the glacier diagnostics, maybe we can discuss if they are necessary for operational runs

- [ ] Fully documented
- [ ] Entry in `whats-new.rst`

pep8speaks commented:
Hello @pat-schmitt! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`oggm/core/flowline.py`](https://github.com/OGGM/oggm/blob/aeae968e2a018dba65d66a163c1f42d1b9444197/oggm/core/flowline.py):

> [Line 3850:64](https://github.com/OGGM/oggm/blob/aeae968e2a018dba65d66a163c1f42d1b9444197/oggm/core/flowline.py#L3850): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 3853:66](https://github.com/OGGM/oggm/blob/aeae968e2a018dba65d66a163c1f42d1b9444197/oggm/core/flowline.py#L3853): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 4012:44](https://github.com/OGGM/oggm/blob/aeae968e2a018dba65d66a163c1f42d1b9444197/oggm/core/flowline.py#L4012): [E126](https://duckduckgo.com/?q=pep8%20E126) continuation line over-indented for hanging indent
> [Line 4028:44](https://github.com/OGGM/oggm/blob/aeae968e2a018dba65d66a163c1f42d1b9444197/oggm/core/flowline.py#L4028): [E126](https://duckduckgo.com/?q=pep8%20E126) continuation line over-indented for hanging indent
> [Line 4103:52](https://github.com/OGGM/oggm/blob/aeae968e2a018dba65d66a163c1f42d1b9444197/oggm/core/flowline.py#L4103): [E126](https://duckduckgo.com/?q=pep8%20E126) continuation line over-indented for hanging indent
> [Line 4186:44](https://github.com/OGGM/oggm/blob/aeae968e2a018dba65d66a163c1f42d1b9444197/oggm/core/flowline.py#L4186): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 4190:44](https://github.com/OGGM/oggm/blob/aeae968e2a018dba65d66a163c1f42d1b9444197/oggm/core/flowline.py#L4190): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 4204:44](https://github.com/OGGM/oggm/blob/aeae968e2a018dba65d66a163c1f42d1b9444197/oggm/core/flowline.py#L4204): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 4211:44](https://github.com/OGGM/oggm/blob/aeae968e2a018dba65d66a163c1f42d1b9444197/oggm/core/flowline.py#L4211): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 4305:64](https://github.com/OGGM/oggm/blob/aeae968e2a018dba65d66a163c1f42d1b9444197/oggm/core/flowline.py#L4305): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 4308:66](https://github.com/OGGM/oggm/blob/aeae968e2a018dba65d66a163c1f42d1b9444197/oggm/core/flowline.py#L4308): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals

* In the file [`oggm/tests/test_models.py`](https://github.com/OGGM/oggm/blob/aeae968e2a018dba65d66a163c1f42d1b9444197/oggm/tests/test_models.py):

> [Line 3238:78](https://github.com/OGGM/oggm/blob/aeae968e2a018dba65d66a163c1f42d1b9444197/oggm/tests/test_models.py#L3238): [E502](https://duckduckgo.com/?q=pep8%20E502) the backslash is redundant between brackets

pat-schmitt commented:
I have now added some glaciers to test the error handling of `run_dynamic_spinup`.
But for this, I added a prepro URL to the `secure_url_retrieve` of `conftest.py`. So as already discussed these glaciers should be combined in a smaller tar file in order to save computing time during the tests.

fmaussion commented on oggm/tests/test_utils.py:
```diff
@@ -1205,6 +1210,24 @@ def test_elev_bands_run(self):
+            dse = dse.load()
+        with xr.open_dataset(sf) as dss:
+            dss = dss.load()
+        assert len(dss.time) > 1
```
@pat-schmitt I rebased your branch and added a commit. You'll need to force pull, but I think I didn't brake anything.

Can you please have a look at this test and what I'm doing wrong? the output only has one year, and I didn't see any error message pointing me to the right direction here

pat-schmitt commented:
Now I have added an option `store_model_evolution` to save the whole model evolution of the spinup and adapted my tests to test if it is working (e.g. delete all files if an error occurs; the model evolution is saved for more than one year if no error occurs; ...).

fmaussion commented:
Thanks a lot @pat-schmitt !
